examples/customresources/demos/complexmodule/moduletest/module_test.go

```diff
@@ -359,3 +359,50 @@ func touchFile(path string) error {
 	}
 	return f.Close()
 }
+
+func TestValidationFailure(t *testing.T) {
+	logger := golog.NewTestLogger(t)
+
+	// bad_modular_validation.json contains a "mybase" modular component that will
+	// fail modular Validation due to a missing "motorL" attribute.
+	cfgFilename, port, err := modifyCfg(
+		utils.ResolveFile("examples/customresources/demos/complexmodule/moduletest/bad_modular_validation.json"), logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, os.Remove(cfgFilename), test.ShouldBeNil)
+	}()
+
+	builder := exec.Command("bash", "-c", "make -s server && cd examples/customresources/demos/complexmodule && go build .")
+	builder.Dir = utils.ResolveFile("")
+	out, err := builder.CombinedOutput()
+	test.That(t, string(out), test.ShouldEqual, "")
+	test.That(t, err, test.ShouldBeNil)
+
+	server := pexec.NewManagedProcess(pexec.ProcessConfig{
+		Name: "bash",
+		Args: []string{"-c", "exec bin/`uname`-`uname -m`/server -config " + cfgFilename},
+		CWD:  utils.ResolveFile("./"),
+		Log:  true,
+	}, logger)
+
+	err = server.Start(context.Background())
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, server.Stop(), test.ShouldBeNil)
+	}()
+
+	rc, err := connect(port, logger)
+	test.That(t, err, test.ShouldBeNil)
+	defer func() {
+		test.That(t, rc.Close(context.Background()), test.ShouldBeNil)
+	}()
+
+	// Assert that motors were added but base was not due to failed Validation.
+	_, err = rc.ResourceByName(motor.Named("motor1"))
+	test.That(t, err, test.ShouldBeNil)
+	_, err = rc.ResourceByName(motor.Named("motor2"))
+	test.That(t, err, test.ShouldBeNil)
+	_, err = rc.ResourceByName(base.Named("base1"))
+	test.That(t, err, test.ShouldNotBeNil)
+	test.That(t, err.Error(), test.ShouldResemble, `resource "rdk:component:base/base1" not found`)
```

Comment:
I'm not sure this is testing the functionality you're adding. Won't the results be the same if the error occurs during building the component, rather than during validation? E.g. either failure cause results in a missing resource here, right?

I'd suggest you check the output of the server to see that it failed with the config validation error you expect as well.

Reply:
Great point; with the new guards in `mybase`, this now passes with or without the changes to `resource_manager.go`. I'll keep this assertion in, as we do want to assert that the base is not built. I've added two assertions to test that a modular validation failure is present in the server output and a build failure is not.